### PR TITLE
Add model - Hierarchical Residual Quantization (HRQVAE)

### PR DESCRIPTION
**Pull request by tomhosking**
This PR adds Hierarchical Residual Quantization (HRQVAE), originally introduced in [this paper](https://aclanthology.org/2022.acl-long.178/) with an implementation [here](https://github.com/tomhosking/hrq-vae). 

It also includes a suggested encoder/decoder for MNIST, which compresses down to a single channel to be quantized. It works fine with the VQVAE encoder/decoder too, though.

I've included tests based on VQVAE, except for the generation pipeline - HRQVAE uses multiple levels of embeddings so it's not so trivial to sample from the continuous space. 

Please let me know what you think!

### CODE REVIEW

**Review by clementchadebec — CHANGES_REQUESTED** [review]
Hello, 

Thank you for this very nice contribution! Happy to see this model included in the codebase. I just reviewed this PR and almost everything looks good to me :). There is still some minor typos that can be addressed before merging in my opinion. In particular, the typing issue mentioned in the comments makes the tests fail. I just tested and the proposed change should do the job since all the tests then pass.

Let me know if you have any questions,

Best,

Clement

File: src/pythae/models/hrq_vae/hrq_vae_model.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        if len(z.shape) == 2:
+            z = z.reshape(z.shape[0], 1, 1, -1)
+
+        z = z.permute(0, 2, 3, 1)
+
+        self.model_config.embedding_dim = z.shape[-1]
+
+        self.quantizer = HierarchicalResidualQuantizer(model_config=model_config)
+
+    def forward(self, inputs: dict[str, Any], **kwargs) -> ModelOutput:
```
[the remark this answers is not in the document]
Comment (tomhosking) [reply src/pythae/models/hrq_vae/hrq_vae_model.py]: Good point, fixed!

File: src/pythae/models/nn/benchmarks/mnist/resnets.py [hunk 2]
```diff
@@ -884,3 +994,121 @@ def forward(self, z: torch.Tensor, output_layer_levels: List[int] = None):
...
+        )
+
+        self.layers = layers
+        self.depth = len(layers)
+
+    def forward(self, z: torch.Tensor, output_layer_levels: List[int] = None):
+        """Forward method
+
+        Args:
+            output_layer_levels (List[int]): The levels of the layers where the outputs are
```
[the remark this answers is not in the document]
Comment (tomhosking) [reply src/pythae/models/nn/benchmarks/mnist/resnets.py:1077]: Done

File: src/pythae/models/nn/benchmarks/mnist/resnets.py [hunk 3]
```diff
@@ -578,6 +578,116 @@ def forward(self, x: torch.Tensor, output_layer_levels: List[int] = None):
...
+        self.layers = layers
+        self.depth = len(layers)
+
+        self.pre_qantized = nn.Conv2d(128, self.latent_dim, 4, 1)
+
+    def forward(self, x: torch.Tensor, output_layer_levels: List[int] = None):
+        """Forward method
+
+        Args:
+            output_layer_levels (List[int]): The levels of the layers where the outputs are
```
[the remark this answers is not in the document]
Comment (tomhosking) [reply src/pythae/models/nn/benchmarks/mnist/resnets.py:652]: Done - the other ResNet models are also missing doc strings for `x`/`z`

File: src/pythae/models/nn/benchmarks/mnist/resnets.py [hunk 4]
```diff
@@ -884,3 +994,121 @@ def forward(self, z: torch.Tensor, output_layer_levels: List[int] = None):
...
         return output
+
+
+class Decoder_ResNet_HRQVAE_MNIST(BaseDecoder):
+    """
+    A ResNet decoder suited for MNIST and Vector Quantized VAE models.
+
+    .. code-block::
+
+        >>> from pythae.models.nn.benchmarks.mnist import Decoder_ResNet_VQVAE_MNIST
```
[the remark this answers is not in the document]
Comment (tomhosking) [reply src/pythae/models/nn/benchmarks/mnist/resnets.py]: Done

File: src/pythae/models/nn/benchmarks/mnist/resnets.py [hunk 5]
```diff
@@ -578,6 +578,116 @@ def forward(self, x: torch.Tensor, output_layer_levels: List[int] = None):
...
+class Encoder_ResNet_HRQVAE_MNIST(BaseEncoder):
+    """
+    A ResNet encoder suited for MNIST and Vector Quantized VAE models.
+
+    It can be built as follows:
+
+    .. code-block::
+
+        >>> from pythae.models.nn.benchmarks.mnist import Encoder_ResNet_VQVAE_MNIST
+        >>> from pythae.models import VQVAEConfig
```
[the remark this answers is not in the document]
Comment (tomhosking) [reply src/pythae/models/nn/benchmarks/mnist/resnets.py]: Good point - done

File: src/pythae/models/hrq_vae/hrq_vae_config.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    Hierarchical Residual Quantization VAE model config config class
+
+    Parameters:
+        input_dim (tuple): The input_data dimension.
+        latent_dim (int): The latent space dimension. Default: None.
+        num_embedding (int): The number of embedding points. Default: 512
+        num_levels (int): Depth of hierarchy. Default: 8
+        kl_weight (float): Weighting of the KL term. Default: 0.0001
+        init_scale (float): Magnitude of the embedding initialisation, should roughly match the encoder. Default: 1.0
+        init_decay_weight (float): Factor by which the magnitude of each successive levels is multiplied. Default: 1.5
```
[the remark this answers is not in the document]
Comment (tomhosking) [reply src/pythae/models/hrq_vae/hrq_vae_config.py:19]: My bad - fixed

File: src/pythae/models/hrq_vae/hrq_vae_config.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    num_embeddings: int = 512
+    num_levels: int = 8
+    kl_weight: float = 0.001
+    init_scale: float = 1.0
+    init_decay_weight: float = 0.5
+    norm_loss_weight: Optional[float] = 0.05
+    norm_loss_scale: float = 1.5
+    temp_schedule_gamma: float = 33.333
+    depth_drop_rate: float = 0.05
```
[the remark this answers is not in the document]
Comment (tomhosking) [reply src/pythae/models/hrq_vae/hrq_vae_config.py]: Done

File: README.md [hunk 8]
```diff
@@ -107,6 +107,7 @@ VAE with Inverse Autoregressive Flows (VAE_IAF) |  [![Open In Colab](https://col
...
 | Regularized AE with gradient penalty (RAE_GP) | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/clementchadebec/benchmark_VAE/blob/main/examples/notebooks/models_training/rae_gp_training.ipynb) | [link](https://arxiv.org/abs/1903.12436) | [link](https://github.com/ParthaEth/Regularized_autoencoders-RAE-/tree/master/) |
 | Riemannian Hamiltonian VAE (RHVAE) | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/clementchadebec/benchmark_VAE/blob/main/examples/notebooks/models_training/rhvae_training.ipynb) | [link](https://arxiv.org/abs/2105.00026) | [link](https://github.com/clementchadebec/pyraug)|
+| Hierarchical Residual Quantization (HRQVAE) | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/clementchadebec/benchmark_VAE/blob/main/examples/notebooks/models_training/hrqvae_training.ipynb) | [link](https://aclanthology.org/2022.acl-long.178/ | [link](https://github.com/tomhosking/hrq-vae)|
```
[the remark this answers is not in the document]
Comment (tomhosking) [reply README.md]: Oops - fixed

**Comment by tomhosking** [comment]
Those fixes should all be in - let me know if you need anything else!

**Review by clementchadebec — APPROVED** [review]
Alright! Thank you for adding those changes in such short notice! Everything looks good now. You can merge whenever ready :)

Thanks again for the contribution! 🚀 

Best,

Clement

**Comment by tomhosking** [comment]
Nice, thanks for the quick follow up! I think you need to do the merge, though :)

**Pull request merged**